_comments-app/components/TextEditor.js

```diff
@@ -0,0 +1,253 @@
+          key={type.label}
+          active={currentStyle.has(type.style)}
+          label={type.label}
+          onToggle={props.onToggle}
+          style={type.style}
+        />
+      )}
+    </div>
+  );
+};
+
+InlineStyleControls.propTypes = {
+  editorState: PropTypes.object.isRequired,
+  onToggle: PropTypes.func.isRequired,
+};
+
+/**
+ * The text editor component.
+ */
+class TextEditor extends Component {
+  constructor(props) {
+    super(props);
+
+    // Initialize state.
+    this.state = { editorState: EditorState.createEmpty() };
+
+    // Bind functions.
+    this.focus = this.focus.bind(this);
+    this.handleEditorChange = this.handleEditorChange.bind(this);
+    this.handleKeyCommand = this.handleKeyCommand.bind(this);
+    this.onTab = this.onTab.bind(this);
+    this.toggleBlockType = this.toggleBlockType.bind(this);
+    this.toggleInlineStyle = this.toggleInlineStyle.bind(this);
+  }
+  focus() {
+    this.editorRef.focus();
+  }
+  handleEditorChange(editorState) {
+    this.setState({editorState: editorState});
+    this.props.setEditorState(this.state.editorState);
+  }
+  handleKeyCommand(command) {
+    const {editorState} = this.state;
+    const newState = RichUtils.handleKeyCommand(editorState, command);
+    if (newState) {
+      this.handleEditorChange(newState);
+      return true;
+    }
+    return false;
+  }
+  onTab(e) {
+    const maxDepth = 4;
+    this.handleEditorChange(RichUtils.onTab(e, this.state.editorState, maxDepth));
+  }
+  toggleBlockType(blockType) {
+    this.handleEditorChange(
+      RichUtils.toggleBlockType(
+        this.state.editorState,
+        blockType
+      )
+    );
+  }
+  toggleInlineStyle(inlineStyle) {
+    this.handleEditorChange(
+      RichUtils.toggleInlineStyle(
+        this.state.editorState,
+        inlineStyle
+      )
+    );
+  }
+  render () {
+    const { editorState } = this.state;
+    let ref = (el) => this.editorRef = el;
+
+    // If user changes block type before entering any text, hide placeholder.
+    let className = 'text-editor__editor';
+    const contentState = editorState.getCurrentContent();
+    if (!contentState.hasText()) {
+      if (contentState.getBlockMap().first().getType() !== 'unstyled') {
+        className += ' text-editor__editor--hide-placeholder';
+      }
+    }
+
+    return(
+      <div className="text-editor">
+        <BlockStyleControls
+          editorState={editorState}
+          onToggle={this.toggleBlockType}
+        />
+        <InlineStyleControls
+          editorState={editorState}
+          onToggle={this.toggleInlineStyle}
+        />
+        <div className={className} onClick={this.focus}>
+          <Editor
+            blockStyleFn={getBlockStyle}
+            customStyleMap={styleMap}
+            editorState={editorState}
+            handleKeyCommand={this.handleKeyCommand}
+            onChange={this.handleEditorChange}
+            onTab={this.onTab}
+            placeholder="What's up?"
```

Comment:
management approved? 😛

Reply:
It's also hard for me to see, when I first look at the comment box, that "What's up" is placeholder text. Some additional styling might be needed to soften the text color? Or maybe it could be removed entirely, but not sure what the best practice is there for accessibility/usability.